File: Python/40.Combination Sum II.py
#找出数组中的数之和为目标数字的所有组合（同一个数不可以使用多次）
#如：[10,1,2,7,6,1,5],8 返回[[1, 1, 6], [1, 2, 5], [1, 7], [2, 6]]
class Solution(object):
    def combinationSum(self, candidates, target):
        candidates = sorted(candidates)
        self.ans, tmp = [], []
        self.dfs(candidates, target, 0, 0, tmp)
        return self.ans

    def dfs(self, candidates, target, p, now, tmp):
        if now == target:
            if not tmp in self.ans:
                self.ans.append(tmp[:])
            return
        for i in range(p, len(candidates)):
            if now + candidates[i] <= target:
                tmp.append(candidates[i])
                self.dfs(candidates, target, i+1, now+candidates[i], tmp)
                tmp.pop()

s = Solution()
print(s.combinationSum([10,1,2,7,6,1,5],8))

File: C++/Backtracking/40.Combination Sum II.cpp
/*
一个数组(无重复元素)，一个目标值，返回所有数组中的元素和为目标值的组合，每个元素只能用一次。
*/
/*
思路：DFS搜索。
*/
#include <iostream>
#include <vector>
#include <algorithm>
using namespace std;

class Solution {
private:
	vector<vector<int>> result;
public:
    vector<vector<int>> combinationSum2(vector<int>& candidates, int target) {
    	sort(candidates.begin(),candidates.end());
    	vector<int> tmp;
        helpCombinationSum(candidates,tmp,0,target);
        return result;
    }
    void helpCombinationSum(vector<int>& candidates,vector<int>& tmp,int start,int target)
    {
    	if(target==0)	result.push_back(tmp),return;
    	for(int i=start; i < candidates.size(); ++i)
    	{
    		if(candidates[i] > target)	return;
    		//i>start是精髓，防止重复，不可以替换成tmp.size()==0，因为重复不一定是第一个。
    		if(i>0 && candidates[i-1]==candidates[i] && i>start)	continue;	
    		tmp.push_back(candidates[i]);
    		helpCombinationSum(candidates,tmp,i+1,target-candidates[i]);
    		tmp.pop_back();
    	}
    }
};
int main()
{
	Solution s;
	vector<int>	candidates{10, 1, 2, 7, 6, 1, 5};
    vector<vector<int>> out = s.combinationSum2(candidates,8);
    for(auto vec:out)
    {
    	for(int n:vec)
    		cout<<n<<" ";
    	cout<<endl;
    }
}

